// @qirolab/nuxt-sanctum-authentication/dist/module.mjs
import { defineNuxtModule, createResolver, useLogger, addPlugin, addImportsDir, addRouteMiddleware } from '@nuxt/kit';
import defu from 'defu';
import { MODULE_NAME, MODULE_CONFIG_KEY } from '../dist/runtime/helpers/module-info.js';

const module = defineNuxtModule({
  meta: {
    name: MODULE_NAME,
    configKey: MODULE_CONFIG_KEY
  },
  defaults: {
    authMode: "cookie",
    userStateKey: "sanctum.authenticated.user",
    token: {
      storageKey: "AUTH_TOKEN",
      provider: "cookie",
      responseKey: "token"
    },
    fetchClientOptions: {
      retryAttempts: false
    },
    csrf: {
      cookieName: "XSRF-TOKEN",
      headerName: "X-XSRF-TOKEN"
    },
    sanctumEndpoints: {
      csrf: "/sanctum/csrf-cookie",
      login: "/login",
      logout: "/logout",
      user: "/api/user"
    },
    redirect: {
      enableIntendedRedirect: false,
      loginPath: "/login",
      guestOnlyRedirect: "/",
      redirectToAfterLogin: "/",
      redirectToAfterLogout: "/"
    },
    middlewareNames: {
      auth: "$auth",
      guest: "$guest"
    },
    logLevel: 3
  },
  setup(_options, _nuxt) {
    const resolver = createResolver(import.meta.url);
    const runtimeDir = resolver.resolve("./runtime");
    _nuxt.options.build.transpile.push(runtimeDir);
    const sanctumOptions = defu(
      _nuxt.options.runtimeConfig.public[MODULE_CONFIG_KEY],
      _options
    );
    _nuxt.options.runtimeConfig.public[MODULE_CONFIG_KEY] = sanctumOptions;
    const logger = useLogger(MODULE_NAME, {
      level: sanctumOptions.logLevel
    });
    logger.start(`Initialize "${MODULE_NAME}" Module...`);
    addPlugin(resolver.resolve("./runtime/plugin"));
    addImportsDir(resolver.resolve("./runtime/composables"));
    addRouteMiddleware({
      name: sanctumOptions.middlewareNames.auth,
      path: resolver.resolve("./runtime/middleware/auth")
    });
    addRouteMiddleware({
      name: sanctumOptions.middlewareNames.guest,
      path: resolver.resolve("./runtime/middleware/guest")
    });
  }
});

export { module as default };
