// @qirolab/nuxt-sanctum-authentication/dist/runtime/services/createFetchService.js
import { useSanctumOptions } from "../composables/useSanctumOptions.js";
import { useTokenStorage } from "../composables/useTokenStorage.js";
import { useCurrentUser } from "../composables/useCurrentUser.js";
import { useCookie, useNuxtApp, useRequestHeaders, useRequestURL } from "#app";
const fetchCsrfCookie = async (config, logger) => {
  try {
    await $fetch(config.sanctumEndpoints.csrf, {
      baseURL: config.apiUrl,
      credentials: "include"
    });
    logger.debug("CSRF cookie has been initialized");
  } catch (error) {
    logger.error("Failed to initialize CSRF cookie", error);
  }
};
const attachCsrfHeader = async (headers, config, logger) => {
  let csrfToken = useCookie(config.csrf.cookieName, { readonly: true });
  if (!csrfToken.value) {
    await fetchCsrfCookie(config, logger);
    csrfToken = useCookie(config.csrf.cookieName, { readonly: true });
  }
  if (!csrfToken.value) {
    logger.warn(
      `${config.csrf.cookieName} cookie is missing, unable to set ${config.csrf.headerName} header`
    );
    return headers ?? {};
  }
  logger.debug(`Added ${config.csrf.headerName} header to the request`);
  return {
    ...headers,
    [config.csrf.headerName]: csrfToken.value
  };
};
const generateServerHeaders = (headers, config) => {
  const clientCookies = useRequestHeaders(["cookie"]);
  const origin = config.appOriginUrl ?? useRequestURL().origin;
  return {
    ...headers,
    Referer: origin,
    Origin: origin,
    ...clientCookies
  };
};
const retrieveAndAttachToken = async (headers, logger) => {
  const nuxtApp = useNuxtApp();
  const token = await useTokenStorage(nuxtApp).get();
  if (!token) {
    logger.debug("Authentication token is not set in the storage");
    return headers;
  }
  return {
    ...headers,
    Authorization: `Bearer ${token}`
  };
};
const processRequestAuth = async (context, config, logger) => {
  const method = context.options.method?.toLowerCase() ?? "get";
  context.options.headers = new Headers({
    Accept: "application/json",
    ...context.options.headers instanceof Headers ? Object.fromEntries(context.options.headers.entries()) : context.options.headers
  });
  if (context.options.body instanceof FormData) {
    context.options.method = "POST";
    context.options.body.append("_method", method.toUpperCase());
  }
  if (config.authMode === "cookie") {
    const SECURE_METHODS = /* @__PURE__ */ new Set(["post", "delete", "put", "patch"]);
    if (import.meta.server) {
      context.options.headers = new Headers(
        generateServerHeaders(
          Object.fromEntries(context.options.headers.entries()),
          config
        )
      );
    }
    if (SECURE_METHODS.has(method)) {
      context.options.headers = new Headers(
        await attachCsrfHeader(
          Object.fromEntries(context.options.headers.entries()),
          config,
          logger
        )
      );
    }
  } else if (config.authMode === "token") {
    context.options.headers = new Headers(
      await retrieveAndAttachToken(
        Object.fromEntries(context.options.headers.entries()),
        logger
      )
    );
  }
};
const getCredentialsMode = () => {
  return "credentials" in Request.prototype ? "include" : void 0;
};
export default function createFetchService(options, logger) {
  const config = useSanctumOptions();
  const httpOptions = {
    baseURL: config.apiUrl,
    credentials: getCredentialsMode(),
    redirect: "manual",
    retry: false,
    onRequest: async (context) => {
      if (options.onRequest) {
        if (Array.isArray(options.onRequest)) {
          for (const hook of options.onRequest) {
            await hook(context);
          }
        } else {
          await options.onRequest(context);
        }
      }
      await processRequestAuth(context, config, logger);
    },
    onResponseError: async (context) => {
      if (context.response.status === 419) {
        logger.warn("CSRF token mismatch");
      } else if (context.response.status === 401) {
        const user = useCurrentUser();
        if (user.value !== null) {
          user.value = null;
        }
      }
      if (options.onResponseError) {
        if (Array.isArray(options.onResponseError)) {
          for (const hook of options.onResponseError) {
            await hook(context);
          }
        } else {
          await options.onResponseError(context);
        }
      }
    }
  };
  return $fetch.create(httpOptions);
}
