// @qirolab/nuxt-sanctum-authentication/dist/runtime/composables/useSanctum.js
import { computed } from "vue";
import { getAuthUser } from "../helpers/get-auth-user.js";
import { extractNestedValue } from "../helpers/utilities.js";
import { useSanctumFetch } from "./useSanctumFetch.js";
import { useSanctumOptions } from "./useSanctumOptions.js";
import { useCurrentUser } from "./useCurrentUser.js";
import { useTokenStorage } from "./useTokenStorage.js";
import { navigateTo, useNuxtApp, useRoute } from "#app";
export const useSanctum = () => {
  const nuxtApp = useNuxtApp();
  const options = useSanctumOptions();
  const user = useCurrentUser();
  const isLoggedIn = computed(() => {
    return user.value !== null;
  });
  async function refreshUser() {
    try {
      user.value = await getAuthUser(useNuxtApp().$sanctumFetch);
    } catch (error) {
      user.value = null;
      console.debug(error);
    }
  }
  async function login(credentials, clientOptions = {}, callback) {
    const { redirect, authMode, sanctumEndpoints } = options;
    const currentRoute = useRoute();
    if (isLoggedIn.value) {
      if (!redirect.redirectToAfterLogin || redirect.redirectToAfterLogin === currentRoute.path) {
        return;
      }
      return await navigateTo(redirect.redirectToAfterLogin);
    }
    const fetchResponse = await useSanctumFetch(
      sanctumEndpoints.login,
      {
        method: "post",
        body: credentials,
        ...clientOptions
      }
    );
    if (authMode === "token") {
      const { token } = options;
      const tokenValue = extractNestedValue(
        fetchResponse,
        token.responseKey
      );
      await useTokenStorage(nuxtApp).set(tokenValue);
    }
    await refreshUser();
    if (callback) {
      return callback(fetchResponse, user.value);
    }
    if (redirect.enableIntendedRedirect) {
      const requestedRoute = currentRoute.query.redirect;
      if (requestedRoute && requestedRoute !== currentRoute.path) {
        return await navigateTo(requestedRoute);
      }
    }
    if (!redirect.redirectToAfterLogin || currentRoute.path === redirect.redirectToAfterLogin) {
      return;
    }
    return await navigateTo(redirect.redirectToAfterLogin);
  }
  async function logout(callback) {
    if (!isLoggedIn.value) {
      return;
    }
    await useSanctumFetch(options.sanctumEndpoints.logout, { method: "post" });
    user.value = null;
    if (options.authMode === "token") {
      await useTokenStorage(nuxtApp).set(void 0);
    }
    if (callback) {
      return callback();
    }
    const currentPath = useRoute().path;
    if (!options.redirect.redirectToAfterLogout || currentPath === options.redirect.redirectToAfterLogout) {
      return;
    }
    await navigateTo(options.redirect.redirectToAfterLogout);
  }
  return {
    options,
    user,
    isLoggedIn,
    refreshUser,
    login,
    logout
  };
};
