// @qirolab/nuxt-sanctum-authentication/dist/runtime/composables/useSanctumForm.js
import { cloneDeep, isEqual, get, set } from "lodash-es";
import { reactive, toRaw, watch } from "vue";
import { hasFile } from "../helpers/has-file.js";
import objectToFormData from "../helpers/object-to-formdata.js";
import { useSanctumFetch } from "./useSanctumFetch.js";
const resolveName = (name) => typeof name !== "string" ? name.target.name : name;
const toSimpleValidationErrors = (errors) => Object.keys(errors).reduce(
  (carry, key) => {
    carry[key] = Array.isArray(errors[key]) ? errors[key][0] : errors[key];
    return carry;
  },
  {}
);
const resolveUrl = (url) => typeof url === "string" ? url : url();
const resolveMethod = (method) => typeof method === "string" ? method.toLowerCase() : method();
const invokeHooks = (hooks, context) => {
  if (Array.isArray(hooks)) {
    hooks.forEach((hook) => hook?.(context));
  } else {
    hooks?.(context);
  }
};
const resolveSubmitOptions = (form, options = {}) => ({
  ...options,
  async onRequest(context) {
    form.processing = true;
    form.setErrors({});
    if ("onRequest" in options && options.onRequest) {
      invokeHooks(options.onRequest, context);
    }
  },
  async onRequestError(context) {
    form.processing = false;
    if ("onRequestError" in options && options.onRequestError) {
      invokeHooks(options.onRequestError, context);
    }
  },
  async onResponse(context) {
    form.processing = false;
    if ("onResponse" in options && options.onResponse) {
      invokeHooks(options.onResponse, context);
    }
  },
  async onResponseError(context) {
    if (context.response.status === 422) {
      form.setErrors(context.response._data.errors);
    }
    if ("onResponseError" in options && options.onResponseError) {
      invokeHooks(options.onResponseError, context);
    }
  }
});
export const useSanctumForm = (method, url, inputs) => {
  const originalData = cloneDeep(inputs);
  const originalInputs = Object.keys(originalData);
  let form = {
    ...cloneDeep(originalData),
    data() {
      const rawData = cloneDeep(toRaw(form));
      return originalInputs.reduce(
        (carry, name) => ({
          ...carry,
          [name]: rawData[name]
        }),
        {}
      );
    },
    setData(data) {
      Object.keys(data).forEach((input) => {
        form[input] = data[input];
      });
      return form;
    },
    processing: false,
    async submit(options) {
      const methodType = resolveMethod(method);
      let preparedData = form.data();
      if (hasFile(preparedData)) {
        preparedData = objectToFormData.serialize(preparedData, {
          indices: true,
          booleansAsIntegers: true
        });
      }
      return useSanctumFetch(resolveUrl(url), {
        method: methodType,
        body: preparedData,
        ...resolveSubmitOptions(form, options)
      });
    },
    errors: {},
    hasErrors: false,
    setErrors(value) {
      const preparedErrors = toSimpleValidationErrors(value);
      if (!isEqual(form.errors, preparedErrors)) {
        form.errors = preparedErrors;
      }
      return form;
    },
    forgetError(name) {
      const newErrors = { ...form.errors };
      Reflect.deleteProperty(newErrors, resolveName(name));
      form.setErrors(newErrors);
      return form;
    },
    invalid(name) {
      return Boolean(form.errors[name]);
    },
    reset(...names) {
      const original = cloneDeep(originalData);
      if (names.length === 0) {
        originalInputs.forEach((name) => {
          form[name] = original[name];
        });
      } else {
        names.forEach((name) => set(form, name, get(original, name)));
      }
      return form;
    }
  };
  form = reactive(form);
  watch(
    () => form.errors,
    () => {
      form.hasErrors = Object.keys(form.errors).length > 0;
    }
  );
  return form;
};
